cli/tools/jupyter.rs

```diff
@@ -0,0 +1,875 @@
+        // TODO: this should be taken from a global,
+        version: "5.3".to_string(),
+      },
+      parent_header: Some(ctx.message.header),
+      session_id: ctx.session_id,
+      metadata: json!({}),
+      content: json!({
+        "execution_state": state.to_string(),
+      }),
+    };
+    // ignore any error when announcing changes
+    let _ = self.iopub_comm.send(msg).await;
+  }
+
+  // TODO(bartlomieju): feels like this info should be a separate struct
+  // instead of KernelMetadata
+  fn get_kernel_info(&self) -> Value {
+    json!({
+      "status": "ok",
+      "protocol_version": self.metadata.protocol_version,
+      "implementation_version": self.metadata.kernel_version,
+      "implementation": self.metadata.implementation_name,
+      "language_info": {
+        "name": self.metadata.language,
+        "version": self.metadata.language_version,
+        "mime": self.metadata.mime,
+        "file_extension": self.metadata.file_ext,
+      },
+      "help_links": [{
+        "text": self.metadata.help_text,
+        "url": self.metadata.help_url
+      }],
+      "banner": self.metadata.banner,
+      "debugger": false
+    })
+  }
+
+  fn get_comm_info() -> Value {
+    json!({
+      "status": "ok",
+      "comms": {}
+    })
+  }
+}
+
+#[derive(Debug)]
+struct KernelMetadata {
+  banner: String,
+  file_ext: String,
+  help_text: String,
+  help_url: String,
+  implementation_name: String,
+  kernel_version: String,
+  language_version: String,
+  language: String,
+  mime: String,
+  protocol_version: String,
+  session_id: String,
+}
+
+impl Default for KernelMetadata {
+  fn default() -> Self {
+    Self {
+      banner: "Welcome to Deno kernel".to_string(),
+      file_ext: ".ts".to_string(),
+      help_text: "<TODO>".to_string(),
+      help_url: "https://github.com/denoland/deno".to_string(),
+      implementation_name: "Deno kernel".to_string(),
+      // FIXME:
+      kernel_version: "0.0.1".to_string(),
+      // FIXME:
+      language_version: "1.16.4".to_string(),
+      language: "typescript".to_string(),
+      // FIXME:
+      mime: "text/x.typescript".to_string(),
+      protocol_version: "5.3".to_string(),
+      session_id: uuid::Uuid::new_v4().to_string(),
+    }
+  }
+}
+
+#[derive(Debug, Deserialize)]
+struct ConnectionSpec {
+  ip: String,
+  transport: String,
+  control_port: u32,
+  shell_port: u32,
+  stdin_port: u32,
+  hb_port: u32,
+  iopub_port: u32,
+  signature_scheme: String,
+  key: String,
+}
+
+struct CommContext {
+  message: Message,
+  hmac: String,
+  session_id: String,
+}
+
+#[derive(Debug, Default, Deserialize, Serialize)]
+struct MessageHeader {
+  msg_id: String,
+  session: String,
+  username: String,
+  date: String,
+  msg_type: String,
+  version: String,
+}
+
+#[derive(Debug)]
+struct Message {
+  is_reply: bool,
+  r#type: String,
+  header: MessageHeader,
+  parent_header: Option<MessageHeader>,
+  metadata: Value,
+  content: Value,
+  session_id: String,
+}
```

Comment:
Feel like this should be split into `Message` and `ReplyMessage`. AFAIK `parent_header` is only present in response messages and for messages received from notebook are always empty dict (`{}`). We could also remove `session_id` and `is_reply` for `Message`. @apowers313 WDYT?

Reply:
`r#type` is also fishy - it's the same value as `header.msg_type`...

Reply:
Refactoring into `RequestMessage` and `ReplyMessage` and creating `From` traits for more more specific types.